storage/remote/queue_manager.go

```diff
@@ -1084,15 +1097,32 @@ func sendWriteRequestWithBackoff(ctx context.Context, cfg config.QueueConfig, l
 
 		backoff = sleepDuration * 2
 
-		if backoff > cfg.MaxBackoff {
-			backoff = cfg.MaxBackoff
+		if backoff > cfg.Retry.MaxBackoff {
+			backoff = cfg.Retry.MaxBackoff
 		}
 
 		try++
 		continue
 	}
 }
 
+func shouldAttempt(policy config.RetryPolicy, minTs int64, retries int) bool {
+	if !policy.IsSet() {
+		return true
+	}
+
+	minSampleAge := timeMilliseconds(time.Now().Add(-time.Duration(policy.MinSampleAge)))
+	if (policy.MinSampleAge != 0 && minTs < minSampleAge) ||
+		(policy.MaxRetries != 0 && retries > policy.MaxRetries) {
+		return false
+	}
+	return true
+}
+
+func timeMilliseconds(t time.Time) int64 {
```

Comment:
promql engine also defines this function, we should consider moving it to the timestamp package if we're going to use it here too